include/utils/fuzzy_equals.h

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Function to check whether both the real and imaginary parts of two complex variables are equal
+ * within an absolute tolerance
+ * @param var1 The first complex variable to be checked
+ * @param var2 The second complex variable to be checked
+ * @param tol The real tolerance to be used for comparing both real and imgainary parts
+ * @return true if both real and imginary parts are equal within tol
+ */
+template <typename T,
+          typename T2,
+          typename T3 = T,
+          typename std::enable_if<ScalarTraits<T3>::value, int>::type = 0>
+bool
+absolute_fuzzy_equals(const std::complex<T> & var1,
+                     const std::complex<T2> & var2,
+                     const T3 & tol = TOLERANCE * TOLERANCE)
+{
+  return absolute_fuzzy_equals(var1.real(), var2.real(), tol) &&
+         absolute_fuzzy_equals(var1.imag(), var2.imag(), tol);
+}
+
+#ifdef LIBMESH_HAVE_METAPHYSICL
+
+#include "metaphysicl/raw_type.h"
+
+/**
+ * Function to check whether two variables are equal within an absolute tolerance
+ * @param var1 The first variable to be checked
+ * @param var2 The second variable to be checked
+ * @param tol The tolerance to be used
+ * @return true if var1 and var2 are equal within tol
+ */
+template <typename T,
+          typename T2,
+          typename T3,
+          typename std::enable_if<ScalarTraits<T>::value && ScalarTraits<T2>::value &&
+                                      ScalarTraits<T3>::value,
+                                  int>::type>
+bool
+absolute_fuzzy_equals(const T & var1, const T2 & var2, const T3 & tol)
+{
+  return (std::abs(MetaPhysicL::raw_value(var1) - MetaPhysicL::raw_value(var2)) <=
+          MetaPhysicL::raw_value(tol));
+}
+
+/**
+ * Function to check whether two variables are equal within a relative tolerance
+ * @param var1 The first variable to be checked
+ * @param var2 The second variable to be checked
+ * @param tol The relative tolerance to be used
+ * @return true if var1 and var2 are equal within relative tol
+ */
+template <typename T, typename T2, typename T3 = Real>
+bool
+relative_fuzzy_equals(const T & var1, const T2 & var2, const T3 & tol = TOLERANCE * TOLERANCE)
+{
+  if constexpr (ScalarTraits<T>::value || TensorTools::MathWrapperTraits<T>::value)
+  {
+    static_assert(TensorTools::TensorTraits<T>::rank == TensorTools::TensorTraits<T2>::rank,
+                  "Mathematical types must be same for arguments to relativelyFuzzEqual");
+    if constexpr (TensorTools::TensorTraits<T>::rank == 0)
+      return absolute_fuzzy_equals(
+          var1,
+          var2,
+          tol * (std::abs(MetaPhysicL::raw_value(var1)) + std::abs(MetaPhysicL::raw_value(var2))));
+    else if constexpr (TensorTools::TensorTraits<T>::rank == 1)
+    {
+      for (const auto i : make_range(libmesh_dim))
+        if (!relative_fuzzy_equals(var1(i), var2(i), tol))
+          return false;
+
```

Comment:
Won't this conclude that the vectors [1e9, 1] and [1e9, 1+1e-5] aren't relatively equal?  That doesn't seem like the behavior we want.

Reply:
The simplest comparable thing to do here is to get the l∞ norm in one pass, then use that as the scale for each difference comparison on a second pass.  We could avoid a 2-pass solution by keeping track of l∞ of both the values and the differences in a single pass and then comparing at the end.  That doesn't let us take advantage of early-return cases, but I don't think I see a way around that.

Reply:
Does the norm we use really matter? I think I'm going to do a bit of consolidation. It looks like we're using the L1 norm in `TypeVector`